sphinx/ext/intersphinx.py

```diff
@@ -410,30 +412,30 @@ def _resolve_reference(env: BuildEnvironment, inv_name: str | None, inventory: I
             if (honor_disabled_refs
                     and (domain_name + ":*") in env.config.intersphinx_disabled_reftypes):
                 continue
-            objtypes = list(domain.object_types)
+            objtypes = domain.object_types.keys()
             res = _resolve_reference_in_domain(env, inv_name, inventory,
                                                honor_disabled_refs,
                                                domain, objtypes,
                                                node, contnode)
             if res is not None:
                 return res
         return None
-    else:
-        domain_name = node.get('refdomain')
-        if not domain_name:
-            # only objects in domains are in the inventory
-            return None
-        if honor_disabled_refs \
-                and (domain_name + ":*") in env.config.intersphinx_disabled_reftypes:
-            return None
-        domain = env.get_domain(domain_name)
-        objtypes = domain.objtypes_for_role(typ)
-        if not objtypes:
-            return None
-        return _resolve_reference_in_domain(env, inv_name, inventory,
-                                            honor_disabled_refs,
-                                            domain, objtypes,
-                                            node, contnode)
+
+    domain_name = node.get('refdomain')
+    if not domain_name:
+        # only objects in domains are in the inventory
+        return None
+    if honor_disabled_refs \
+            and (domain_name + ":*") in env.config.intersphinx_disabled_reftypes:
+        return None
+    domain = env.get_domain(domain_name)
+    objtypes = {}.fromkeys(domain.objtypes_for_role(typ)).keys()
```

Comment:
We could defer the creation of the dictionary:

```python
_objtypes = domain.objtypes_for_role(typ)
if not _objtypes:
    return None
objtypes = {}.fromkeys(_objtypes).keys()
```

Reply:
This was included in https://github.com/sphinx-doc/sphinx/pull/11338/commits/6a9e617a49041b042d1666ba49a726f06aa46601, if I understand you correctly!

A